File: problems/368_largest_divisible_subseet.java
/*
 * Given a set of distinct positive integers nums, return the largest subset answer such that every pair (answer[i], answer[j]) of elements in this subset satisfies:

answer[i] % answer[j] == 0, or
answer[j] % answer[i] == 0
If there are multiple solutions, return any of them.



Example 1:

Input: nums = [1,2,3]
Output: [1,2]
Explanation: [1,3] is also accepted.
Example 2:

Input: nums = [1,2,4,8]
Output: [1,2,4,8]


Constraints:

1 <= nums.length <= 1000
1 <= nums[i] <= 2 * 109
All the integers in nums are unique.
*/

class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int n = nums.length;
        if (n == 0) return new ArrayList();
        List<ArrayList> EDS = new ArrayList();
        for (int num : nums) EDS.add(new ArrayList());
        Arrays.sort(nums);
        for (int i = 0; i < n; ++i) {
        List<Integer> maxSubset = new ArrayList();
        for (int k = 0; k < i; ++k)
            if (nums[i] % nums[k] == 0 && maxSubset.size() < EDS.get(k).size())
            maxSubset = EDS.get(k);
        EDS.get(i).addAll(maxSubset);
        EDS.get(i).add(nums[i]);
        }
        List<Integer> ret = new ArrayList();
        for (int i = 0; i < n; ++i)
        if (ret.size() < EDS.get(i).size()) ret = EDS.get(i);
        return ret;
    }
}

File: problems/368_largest_divisible_subseet.py
'''
Given a set of distinct positive integers, find the largest subset such that every pair (Si, Sj) of elements in this subset satisfies:

Si % Sj = 0 or Sj % Si = 0.

If there are multiple solutions, return any subset is fine.

Example 1:

Input: [1,2,3]
Output: [1,2] (of course, [1,3] will also be ok)
Example 2:

Input: [1,2,4,8]
Output: [1,2,4,8]
'''

# O(N^2) time and O(N^2) space
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        subsets = {-1: set()}
        for num in sorted(nums):
            subsets[num] = max([subsets[k] for k in subsets if num % k == 0], key=len) | {num}
            
        return list(max(subsets.values(), key=len))
        

